### DISCUSSION TITLE
How to build an extension without including the external libs?

### DISCUSSION THREAD

**Discussion by mattiaz9** [directus/directus#8143] [Archive]
I was used to build my extensions with rollup, but now I see we're supposed to use `npx directus-extension build` to build an extension.

The problem I see is that my extensions went from few kb to 200+kb each because I think they include dependencies like vue,  @directus/shared/composables, vue-i18n etc.

Those dependencies shouldn't be included in each extensions right?

Also I have 2 more questions about this new builder:
- Does it have a 'watch' option?
- Can it build multiple extensions from the root directory of the project?

**Comment by joselcvarela** [comment]
@nickrum can you help?

**Comment by mattiaz9** [comment]
After some research I understood why the bundle was larger, it was because I was importing from `@directus/shared/composables` instead of `@directus/extensions-sdk`. That was my bad, but still we need some better documentation on how to build extensions.

Regarding the new builder I saw it has a watch option, but there is no way to have multiple input files. That means if someone have, for instance, 10 extensions it would also have to install `node_modules` in 10 subdirectories. I think one `node_modules` is enough :)

At the end I managed to configure rollup properly so I'm sticking with it for now.

**Reply by nickrum** [reply]
Yes, Directus-specific utilities have to be imported from `@directus/extensions-sdk`. If you import directly from `@directus/shared/composables` and externalize this import, your extension won't work because Directus (or rather your browser) wouldn't know where to find this import. But, I fully agree that the documentation is currently lacking.

If your main concern is module duplication, you can just use npm's workspaces feature. Building multiple extensions at once is already planned, but it will likely involve creating separate packages for each extension.

Using your own rollup config is a prefectly fine solution. Glad it worked for you in the end :)

**Comment by nickrum** [comment]
> The problem I see is that my extensions went from few kb to 200+kb each because I think they include dependencies like vue, @directus/shared/composables, vue-i18n etc.

Quite the opposite is true. One of the reasons for building the extensions-sdk in the first place was to be able to easily externalize those dependencies shared between the Directus App and extensions.
